src/Mvc/src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs

```diff
@@ -172,18 +167,11 @@ public string DisplayNameFor<TResult>(Expression<Func<TModel, TResult>> expressi
                 throw new ArgumentNullException(nameof(expression));
             }
 
-            var modelExplorer = ExpressionMetadataProvider.FromLambdaExpression<TModelItem, TResult>(
-                expression,
+            var modelExpression = _modelExpressionProvider.CreateModelExpression(
                 new ViewDataDictionary<TModelItem>(ViewData, model: null),
-                MetadataProvider);
-
-            var expressionText = ExpressionHelper.GetExpressionText(expression, _expressionTextCache);
-            if (modelExplorer == null)
```

Comment:
Should keep this check (using `modelExplorer.ModelExplorer`) because the explorer will be `null` if the helper can't compile the expression.